### PR TITLE
test: add some test for rebalance

### PR DESCRIPTION
**Pull request by Gala-R**

### CODE REVIEW

File: controllers/apps/v1beta4/rebalance_controller_test.go [hunk 1]
```diff
@@ -63,3 +69,197 @@ func TestGetRebalanceStatus(t *testing.T) {
...
+		assert.ErrorContains(t, err, "nothing_to_balance")
+	})
+
+	t.Run("check requestAPI start rebalance success", func(t *testing.T) {
+		f.requestAPI = func(method, path string, body []byte) (resp *http.Response, respBody []byte, err error) {
+			return &http.Response{StatusCode: http.StatusOK}, []byte(`{"data":[],"code":0}`), nil
+		}
+		err := startRebalance(f, rebalance, emqx, emqxNodeName)
+		assert.Nil(t, err)
+	})
```
Comment (Rory-Z) [inline controllers/apps/v1beta4/rebalance_controller_test.go:148]: Looks it repeat from `t.Run("check requestAPI args", func(t *testing.T) {})`

File: controllers/apps/v1beta4/rebalance_controller_test.go [hunk 2]
```diff
@@ -63,3 +69,197 @@ func TestGetRebalanceStatus(t *testing.T) {
...
+		assert.ErrorContains(t, err, "rebalance is disabled")
+	})
+
+	t.Run("check requestAPI stop rebalance success", func(t *testing.T) {
+		f.requestAPI = func(method, path string, body []byte) (resp *http.Response, respBody []byte, err error) {
+			return &http.Response{StatusCode: http.StatusOK}, []byte(`{"data":[],"code":0}`), nil
+		}
+		err := stopRebalance(f, rebalance)
+		assert.Nil(t, err)
+	})
```
Comment (Rory-Z) [inline controllers/apps/v1beta4/rebalance_controller_test.go:231]: Looks it repeat from t.Run("check requestAPI args", func(t *testing.T) {})

File: controllers/apps/v1beta4/rebalance_controller_test.go [hunk 3]
```diff
@@ -63,3 +69,197 @@ func TestGetRebalanceStatus(t *testing.T) {
...
+	t.Run("check requestAPI return error", func(t *testing.T) {
+		f.requestAPI = func(method, path string, body []byte) (resp *http.Response, respBody []byte, err error) {
+			return nil, nil, errors.New("fake error")
+		}
+		err := stopRebalance(f, rebalance)
+		assert.Error(t, err, "fake error")
+	})
+}
+
+func getRequestBody(rebalance *appsv1beta4.Rebalance, nodes []string) []byte {
```
Comment (Rory-Z) [inline controllers/apps/v1beta4/rebalance_controller_test.go:242]: What's this ?

File: controllers/apps/v1beta4/rebalance_controller_test.go [hunk 4]
```diff
@@ -63,3 +69,197 @@ func TestGetRebalanceStatus(t *testing.T) {
 	})
 
 }
+
+func TestStartRebalance(t *testing.T) {
+	f := &fakePW{}
+	rebalance := &appsv1beta4.Rebalance{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "rebalance-test",
+		},
```
Comment (Rory-Z) [inline controllers/apps/v1beta4/rebalance_controller_test.go:78]: This is needless

File: controllers/apps/v1beta4/rebalance_controller_test.go [hunk 5]
```diff
@@ -63,3 +69,197 @@ func TestGetRebalanceStatus(t *testing.T) {
...
+		assert.Error(t, err, "fake error")
+	})
+}
+
+func TestStopRebalance(t *testing.T) {
+	f := &fakePW{}
+	rebalance := &appsv1beta4.Rebalance{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "rebalance-test",
+		},
```
Comment (Rory-Z) [inline controllers/apps/v1beta4/rebalance_controller_test.go:164]: This is needless

File: controllers/apps/v1beta4/rebalance_controller_test.go [hunk 6]
```diff
@@ -63,3 +69,197 @@ func TestGetRebalanceStatus(t *testing.T) {
...
+				WaitHealthCheck:  10,
+				WaitTakeover:     10,
+				AbsSessThreshold: 10,
+				AbsConnThreshold: 10,
+				RelConnThreshold: "1.1",
+				RelSessThreshold: "1.1",
+			},
+		},
+		Status: appsv1beta4.RebalanceStatus{
+			Phase: "Processing",
```
Comment (Rory-Z) [inline controllers/apps/v1beta4/rebalance_controller_test.go:179]: This line is needless

File: controllers/apps/v1beta4/rebalance_controller_test.go (hunk 5 above)
Comment (Gala-R) [reply controllers/apps/v1beta4/rebalance_controller_test.go:164]: > This is needless

delete

File: controllers/apps/v1beta4/rebalance_controller_test.go (hunk 4 above)
Comment (Gala-R) [reply controllers/apps/v1beta4/rebalance_controller_test.go:78]: > This is needless

delete

File: controllers/apps/v1beta4/rebalance_controller_test.go (hunk 6 above)
Comment (Gala-R) [reply controllers/apps/v1beta4/rebalance_controller_test.go:179]: > This line is needless

why？ To stop rebalance, you need to confirm whether rebalance is in progress

File: controllers/apps/v1beta4/rebalance_controller_test.go (hunk 2 above)
Comment (Gala-R) [reply controllers/apps/v1beta4/rebalance_controller_test.go:231]: > Looks it repeat from t.Run("check requestAPI args", func(t *testing.T) {})

Ok, i got it

File: controllers/apps/v1beta4/rebalance_controller_test.go (hunk 1 above)
Comment (Gala-R) [reply controllers/apps/v1beta4/rebalance_controller_test.go:148]: > Looks it repeat from `t.Run("check requestAPI args", func(t *testing.T) {})`

ok, i got it

File: controllers/apps/v1beta4/rebalance_controller_test.go (hunk 3 above)
Comment (Gala-R) [reply controllers/apps/v1beta4/rebalance_controller_test.go:242]: > What's this ?

move this code in test case

File: controllers/apps/v1beta4/rebalance_controller_test.go [hunk 7]
```diff
@@ -63,3 +68,237 @@ func TestGetRebalanceStatus(t *testing.T) {
...
+				ConnEvictRate:    5,
+				SessEvictRate:    5,
+				WaitHealthCheck:  10,
+				WaitTakeover:     10,
+				AbsSessThreshold: 10,
+				AbsConnThreshold: 10,
+				RelConnThreshold: "1.1",
+				RelSessThreshold: "1.1",
+			},
+		},
```
Comment (Rory-Z) [inline controllers/apps/v1beta4/rebalance_controller_test.go:162]: Looks this is needless ?

Comment (Gala-R) [reply controllers/apps/v1beta4/rebalance_controller_test.go:162]: > 

why ?

**Review by Rory-Z — APPROVED** [review]

**Pull request merged**